File: tools/extract_docx.py
#!/usr/bin/env python3
"""Extract a .docx to greppable, diffable text WITHOUT losing table structure.

Usage:  python tools/extract_docx.py <input.docx> <output.md>
Stdlib only. No pip install.
"""
import html
import re
import sys
import zipfile

CELL = "\x00CELL\x00"
ROW = "\x00ROW\x00"


def xml_to_text(xml: str) -> str:
    # Structure markers must be inserted BEFORE tags are stripped.
    xml = re.sub(r"<w:br[^>]*/>", "\n", xml)
    xml = re.sub(r"<w:tab[^>]*/>", "\t", xml)
    xml = re.sub(r"</w:p>", "\n", xml)
    xml = re.sub(r"</w:tc>", CELL, xml)
    xml = re.sub(r"<w:tr[ >]", ROW + r"\g<0>", xml)  # close the preceding paragraph
    xml = re.sub(r"</w:tr>", ROW, xml)
    text = re.sub(r"<[^>]+>", "", xml)
    text = html.unescape(text)

    # Rebuild tables as pipe-delimited rows.
    out = []
    for row in text.split(ROW):
        if CELL in row:
            cells = [c.replace("\n", " ").strip() for c in row.split(CELL)]
            if cells and cells[-1].strip() == "":
                cells.pop()
            out.append("| " + " | ".join(cells) + " |")
        else:
            out.append(row)
    text = "\n".join(out)

    text = re.sub(r"[ \t]+\n", "\n", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    return text.strip() + "\n"


def main(src: str, out: str) -> None:
    parts, missing = [], []
    with zipfile.ZipFile(src) as z:
        names = set(z.namelist())
        for member, label in [
            ("word/document.xml", None),
            ("word/footnotes.xml", "FOOTNOTES"),
            ("word/endnotes.xml", "ENDNOTES"),
        ]:
            if member not in names:
                missing.append(member)
                continue
            body = xml_to_text(z.read(member).decode("utf-8", "replace"))
            if label:
                body = f"\n\n---\n\n# {label}\n\n{body}"
            parts.append(body)

    text = "".join(parts)
    with open(out, "w", encoding="utf-8") as fh:
        fh.write(text)

    ids = re.findall(r"\b((?:CP|KPP|TPP|ASM|SR|PR|OI)-[A-Z0-9]+(?:-[0-9]+[a-z]?)?)", text)
    uniq = sorted(set(ids))
    fams = sorted({i.rsplit("-", 1)[0] for i in uniq})
    print(f"wrote {out}")
    print(f"  chars           {len(text):,}")
    print(f"  lines           {text.count(chr(10)):,}")
    print(f"  table rows      {text.count('|' + chr(10)):,}")
    print(f"  identifiers     {len(ids):,} total / {len(uniq):,} unique")
    print(f"  families ({len(fams)}) {', '.join(fams)}")
    if missing:
        print(f"  absent parts    {', '.join(missing)}")
    if not uniq:
        print("  !! NO CP-/KPP-/TPP- IDENTIFIERS FOUND — stop and inspect before continuing.")


if __name__ == "__main__":
    if len(sys.argv) != 3:
        sys.exit(__doc__)
    main(sys.argv[1], sys.argv[2])

File: tools/extract_docx.mjs
#!/usr/bin/env node
/**
 * Extract a .docx to greppable, diffable text WITHOUT losing table structure.
 *
 * Usage:  node tools/extract_docx.mjs <input.docx> <output.md>
 * Node stdlib only (zlib + fs). No npm install, no Python.
 *
 * R131 [S2]: this began as a port of tools/extract_docx.py, which was committed
 * having never been run, so nothing had ever checked the port against its
 * original. Running both on all three .docx files in the repo settled it: the
 * content is byte for byte identical, and the only divergence is line endings -
 * Python's text-mode write translates 
 to 
 on Windows, this writes 
.
 * On the framework document that is 1,614,507 bytes against 1,598,905, equal
 * after normalising newlines; the identifier, line and table-row counts agree
 * exactly on every file. The Python original is deleted rather than kept as a
 * second implementation of one job: package.json pins Node 22 through Volta, so
 * Node is the runtime the repo can assume. The two generators that need
 * python-docx remain Python; which runtime each tool needs is declared in
 * src/lib/toolchain-check.ts and gated by a self-test.
 */
import { readFileSync, writeFileSync } from "node:fs";
import { inflateRawSync } from "node:zlib";

const CELL = "\x00CELL\x00";
const ROW = "\x00ROW\x00";

/** Minimal ZIP reader: central directory walk, stored or deflated members. */
function readZip(path) {
  const buf = readFileSync(path);
  let eocd = -1;
  for (let i = buf.length - 22; i >= 0; i--) {
    if (buf.readUInt32LE(i) === 0x06054b50) { eocd = i; break; }
  }
  if (eocd < 0) throw new Error(`not a zip file: ${path}`);
  const count = buf.readUInt16LE(eocd + 10);
  let off = buf.readUInt32LE(eocd + 16);
  const members = new Map();
  for (let n = 0; n < count; n++) {
    if (buf.readUInt32LE(off) !== 0x02014b50) throw new Error("bad central directory");
    const method = buf.readUInt16LE(off + 10);
    const compSize = buf.readUInt32LE(off + 20);
    const nameLen = buf.readUInt16LE(off + 28);
    const extraLen = buf.readUInt16LE(off + 30);
    const commentLen = buf.readUInt16LE(off + 32);
    const localOff = buf.readUInt32LE(off + 42);
    const name = buf.toString("utf8", off + 46, off + 46 + nameLen);
    const lNameLen = buf.readUInt16LE(localOff + 26);
    const lExtraLen = buf.readUInt16LE(localOff + 28);
    const dataStart = localOff + 30 + lNameLen + lExtraLen;
    const raw = buf.subarray(dataStart, dataStart + compSize);
    members.set(name, method === 0 ? raw : inflateRawSync(raw));
    off += 46 + nameLen + extraLen + commentLen;
  }
  return members;
}

const ENTITIES = { amp: "&", lt: "<", gt: ">", quot: '"', apos: "'", "#39": "'" };
function unescapeHtml(s) {
  return s.replace(/&(#x[0-9a-fA-F]+|#[0-9]+|[a-zA-Z]+);/g, (m, ref) => {
    if (ref[0] === "#") {
      const cp = ref[1] === "x" || ref[1] === "X"
        ? parseInt(ref.slice(2), 16)
        : parseInt(ref.slice(1), 10);
      return Number.isFinite(cp) ? String.fromCodePoint(cp) : m;
    }
    const lower = ref.toLowerCase();
    return Object.prototype.hasOwnProperty.call(ENTITIES, lower) ? ENTITIES[lower] : m;
  });
}

function xmlToText(xml) {
  // Structure markers must be inserted BEFORE tags are stripped.
  xml = xml.replace(/<w:br[^>]*\/>/g, "\n");
  xml = xml.replace(/<w:tab[^>]*\/>/g, "\t");
  xml = xml.replace(/<\/w:p>/g, "\n");
  xml = xml.replace(/<\/w:tc>/g, CELL);
  xml = xml.replace(/<w:tr[ >]/g, ROW + "$&"); // close the preceding paragraph
  xml = xml.replace(/<\/w:tr>/g, ROW);
  let text = xml.replace(/<[^>]+>/g, "");
  text = unescapeHtml(text);

  // Rebuild tables as pipe-delimited rows.
  const out = [];
  for (const row of text.split(ROW)) {
    if (row.includes(CELL)) {
      const cells = row.split(CELL).map((c) => c.replace(/\n/g, " ").trim());
      if (cells.length && cells[cells.length - 1].trim() === "") cells.pop();
      out.push("| " + cells.join(" | ") + " |");
    } else {
      out.push(row);
    }
  }
  text = out.join("\n");

  text = text.replace(/[ \t]+\n/g, "\n");
  text = text.replace(/\n{3,}/g, "\n\n");
  return text.trim() + "\n";
}

function main(src, out) {
  const zip = readZip(src);
  const parts = [];
  const missing = [];
  for (const [member, label] of [
    ["word/document.xml", null],
    ["word/footnotes.xml", "FOOTNOTES"],
    ["word/endnotes.xml", "ENDNOTES"],
  ]) {
    if (!zip.has(member)) { missing.push(member); continue; }
    let body = xmlToText(zip.get(member).toString("utf8"));
    if (label) body = `\n\n---\n\n# ${label}\n\n${body}`;
    parts.push(body);
  }

  const text = parts.join("");
  writeFileSync(out, text, "utf8");

  const ids = text.match(/\b(?:CP|KPP|TPP|ASM|SR|PR|OI)-[A-Z0-9]+(?:-[0-9]+[a-z]?)?/g) || [];
  const uniq = [...new Set(ids)].sort();
  const fams = [...new Set(uniq.map((i) => i.slice(0, i.lastIndexOf("-"))))].sort();
  const n = (v) => v.toLocaleString("en-US");
  const lines = (text.match(/\n/g) || []).length;
  const rows = (text.match(/\|\n/g) || []).length;
  console.log(`wrote ${out}`);
  console.log(`  chars           ${n(text.length)}`);
  console.log(`  lines           ${n(lines)}`);
  console.log(`  table rows      ${n(rows)}`);
  console.log(`  identifiers     ${n(ids.length)} total / ${n(uniq.length)} unique`);
  console.log(`  families (${fams.length}) ${fams.join(", ")}`);
  if (missing.length) console.log(`  absent parts    ${missing.join(", ")}`);
  if (!uniq.length) {
    console.log("  !! NO CP-/KPP-/TPP- IDENTIFIERS FOUND - stop and inspect before continuing.");
  }
}

const [, , src, out] = process.argv;
if (!src || !out) {
  console.error("Usage: node tools/extract_docx.mjs <input.docx> <output.md>");
  process.exit(1);
}
main(src, out);
